### ISSUE TITLE
`npm install -g` a package with peerDependency fails with "Could not determine the name of the package that was just installed."

### ISSUE THREAD

**Issue by cspotcode** [volta-cli/volta#996]
I have created this minimal reproduction:

https://gist.github.com/cspotcode/c07208f11ed73dc7f81a0200e5aeb797

However, I first encountered this issue trying to do `npm install -g ts-node`

I'm using npm v7 which automatically installs `peerDependencies`.

Is volta observing that npm has added multiple `node_modules` and unable to figure out which is the peer?

**Comment by charlespierce** [comment]
@cspotcode Thanks for reporting this, and for the easy reproduction! I've done a bit of investigation into the automatic peer dependency behavior. At its core, I think that behavior is a bit at odds with Volta's sandboxing of individual packages. The potential concern I have, using `ts-node` as an example, is that with `typescript` installed automatically in the same workspace as `ts-node`, there is no way (currently) for Volta users to then later update the TypeScript version that is associated with `ts-node`. Installing a new version of TypeScript globally would do so in a separate sandbox and while there is a mechanism in place to link the sandboxes, it has a lower precedence than the automatically installed version, so it would never use the new one.

As a result, until / unless we can come up with a better model for handling the global packages, I'm inclined to take advantage of the [`legacy-peer-deps`](https://docs.npmjs.com/cli/v7/using-npm/config#legacy-peer-deps) option to disable the automatic installation (for global installs _only_). That's not ideal—and not a recommended option per npm's own docs—but it will allow the global installs to continue to work as they always have.

**Comment by cspotcode** [comment]
Interesting, to ensure that I understand completely, to install `ts-node` globally I would do this?

```
npm install -g ts-node
npm install -g typescript@<desired-version>
npm install -g @types/node@<desired-version>

# The commands cannot be combined like this:
npm install -g ts-node typescript@<desired-version> @types/node@<desired-version>
```

That works for me.

I'm unfamiliar with volta's sandboxes, but does it want to allow the globally-installed `tsc` version to be *different* than the `typescript` peer available to `ts-node`?

**Comment by charlespierce** [comment]
> I'm unfamiliar with volta's sandboxes, but does it want to allow the globally-installed `tsc` version to be _different_ than the `typescript` peer available to `ts-node`? EDIT clarificiation: is that a use-case that volta wants to support?

No, actually Volta wants the exact opposite: We want the user to have control over the `typescript` peer. With the automatic peer dependency installation, we actually would have two different versions: The one you installed globally and the one that was automatically installed via `peerDependencies`. As a (simplified) example, this is what the file structure would look like _without_ automatic peers:

```
.
├── ts-node-volta
│   └── node_modules
│       └── ts-node
│           └── package.json
└── typescript-volta
    └── node_modules
        └── typescript
            └── package.json
```

With some `NODE_PATH` magic that allows the two top-level packages (with `-volta` to see one another. However, when peers are automatically installed, it would look more like this:

```
.
├── ts-node-volta
│   └── node_modules
│       ├── ts-node
│       │   └── package.json
│       └── typescript
│           └── package.json
└── typescript-volta
    └── node_modules
        └── typescript
            └── package.json
```

The current interface only allows you to interact with the top level `-volta` directories (effectively), however the Node module resolution algorithm will always find the internal `typescript` when calling it from `ts-node`. So there would be no way to affect which version was installed with the automatic peer installation.

At least that's the current state of things; there may be a way to redesign the layouts to allow for more fine-grained control.

> Interesting, to ensure that I understand completely, to install `ts-node` globally I would do this?
> 
> ```
> npm install -g ts-node
> npm install -g typescript@<desired-version>
> npm install -g @types/node@<desired-version>
> 
> # The commands cannot be combined like this:
> npm install -g ts-node typescript@<desired-version> @types/node@<desired-version>
> ```
> 
> That works for me.

With the current Volta release and using npm < 7 (or turning off automatic peer dependencies), either of those commands will work. Volta does some behind-the-scenes work to effectively turn the 2nd of those commands into the first one transparently.

**Comment by cspotcode** [comment]
Ok, thanks.

Do you need to support cases where someone wants to install ts-node with typescript 3.9, and install tsc 4.0?  Letting the user control the peer version *and* letting it be different than the global tsc version?

I have no idea if this use-case is worth considering or not; I assume volta's team has given this more thought than I have.

**Comment by charlespierce** [comment]
@cspotcode I don't think we have a need to support that use-case, though I'm happy to be convinced otherwise. Our general philosophy is to make the behavior as close as possible to how it would work without Volta, to reduce surprise as much as possible. In a non-Volta situation with Node installed directly, you wouldn't be able to have a global typescript separate from the one that global ts-node uses, so to first approximation we would want to mirror that behavior.

I could see an approach where you can define "groups" of packages that are installed together. I think we may even have talked a bit about it when designing the package model. A major blocker for that, however, is that it can get really complicated to visualize and understand what you're installing / upgrading and where. So that's another reason why we wanted to use a one-level model: Either you have something installed or you don't, there isn't additional complication of _where_ you have something installed.

**Comment by cspotcode** [comment]
@charlespierce Ok, cool, that all makes sense to me.  I don't personally need to pick multiple versions.
